binary_search_tree/tree.py

```diff
@@ -1,59 +1,147 @@
 class TreeNode:
-    def __init__(self, key, val = None):
+    def __init__(self, key, val=None):
         if val == None:
             val = key
 
         self.key = key
         self.value = val
         self.left = None
         self.right = None
-        
+
+    def height(self):
+        if self.right and self.left:
+            return 1 + max(self.left.height(), self.right.height())
+        elif self.left:
+            return 1 + self.left.height()
+        elif self.right:
+            return 1 + self.right.height()
+        else:
+            return 1
 
 
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
-    def add(self, key, value = None):
-        pass
+    # Time Complexity: O(log N)
+    # Space Complexity: O(1)
+
+    def add_helper(self, current_node, key, value):
```

Comment:
The space complexity is O(log n) if the tree is balanced due to the recursive call stack.